examples/social/src/state.rs

```diff
@@ -12,4 +15,6 @@ pub struct SocialState {
     pub own_posts: LogView<OwnPost>,
     /// Posts we received from authors we subscribed to.
     pub received_posts: CustomMapView<Key, Post>,
+    /// Other chains events we have already processed.
+    pub processed_events: MapView<ChainId, u32>,
```

Comment:
Normally this kind of tracking would be done in service, not in the contract - right? Otherwise the indexing over all emitted events in `process_stream` is unnecessarily expensive.

Reply:
It has to be the contract that keeps this map: When `process_streams` gets called, it needs to know which new events to read.

Reply:
I thought this decision - which events to read - will be done externally as well. Otherwise contract execution pays for detecting what to ignore and ignoring that.

Reply:
@afck The system has to maintain a similar map for deduplication purposes, right? So what we could do is pass the previous stream index:
```
async fn process_streams(&mut self, streams: Vec<(ChainId, StreamId, /*previous*/ Option<u32>, /*new*/ u32)>
```

Reply:
Btw. we should probably define something like `struct StreamUpdate { chain_id, stream_id, previous_index, new_index }`.

Reply:
(and ofc we can also use `previous_index_plus_one` aka `previous_next_index` to avoid the `Option`)

Reply:
Based on the `UpdateStreams` handling in `system.rs`, it looks like this should be easy. I'm excited :)

Reply:
This way, contracts have the freedom to handle every element of a stream exactly once -- or just the last one whenever a notification arrives. In previous designs, it was hard to support both modes efficiently.

Reply:
> The system has to maintain a similar map for deduplication purposes, right?

Not really, no. But I can add that. I agree that would be more user-friendly.

Reply:
oh I see. `EventSubscriptions::next_index` is a bit different